src/FromData.purs

```diff
@@ -1,31 +1,142 @@
 module FromData
   ( class FromData
+  , class FromDataArgs
+  , class FromDataWithIndex
+  , class FromDataArgsRL
+  , fromDataArgsRec
   , fromData
+  , fromDataArgs
+  , fromDataWithIndex
+  , genericFromData
   ) where
 
 import Prelude
 
+import ConstrIndex (class HasConstrIndex, constrIndex)
+import Control.Alternative ((<|>), guard)
+import Data.Array (uncons)
 import Data.Array as Array
 import Data.BigInt (BigInt)
+import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
+import Data.Generic.Rep as G
 import Data.List (List)
 import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(Nothing, Just), maybe)
 import Data.Ratio (Ratio, reduce)
+import Data.Symbol (class IsSymbol, reflectSymbol)
 import Data.Traversable (for, traverse)
 import Data.Tuple (Tuple(Tuple))
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Data.Unfoldable (class Unfoldable)
 import Helpers (bigIntToUInt)
+import Prim.Row as Row
+import Prim.RowList as RL
 import Prim.TypeError (class Fail, Text)
+import Record as Record
+import Type.Proxy (Proxy(Proxy))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Bytes, Constr, List, Map, Integer))
 
+-- | Classes
+
+class FromData :: Type -> Constraint
 class FromData (a :: Type) where
   fromData :: PlutusData -> Maybe a
 
+class FromDataWithIndex :: Type -> Type -> Constraint
+class FromDataWithIndex a ci where
```

Comment:
What are your thoughts on writing as
```
class HasConstrIndex ci <= FromDataWithIndex a ci where
  fromDataWithIndex :: Proxy a -> Proxy ci -> PlutusData -> Maybe a
 ```
 Then the constraint is explicit in the typeclass, rather than implicit in the method?
In this case `instance (FromDataWithIndex l a, FromDataWithIndex r a) => FromDataWithIndex (G.Sum l r) a where` doesn't require the extra constraint by superclass constraint as opposed to method constraint.  See my comments on `ToDataWithIndex (G.Sum l r)`.
  
  This of course means
  ```
  instance (IsSymbol n, HasConstrIndex a, FromDataArgs arg) => FromDataWithIndex (G.Constructor n arg) a where
  ```
  notice extra constraint, but it is now **symmetric** with `ToDataWithIndex (G Constructor ...)`